### PR TITLE
✨ CS-Flow model

### LINKED ISSUE #610: Add CS-Flow Model
### Discussed in https://github.com/openvinotoolkit/anomalib/discussions/575

<div type='discussions-op-text'>

<sup>Originally posted by **djdameln** April 22, 2022</sup>
Add implementation of the CS-Flow algorithm.
See https://paperswithcode.com/paper/fully-convolutional-cross-scale-flows-for</div>

### PR DESCRIPTION
**Pull request by ashwinvaidya17**
# Description

- Adds CS-Flow
- Use FrEIA from pip
- Add feature extraction using TorchFX

- Fixes #610 

## Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Refactor (non-breaking change which refactors the code base)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [x] My code follows the [pre-commit style and check guidelines](https://openvinotoolkit.github.io/anomalib/guides/using_pre_commit.html#pre-commit-hooks) of this project.
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing tests pass locally with my changes

### CODE REVIEW

File: anomalib/models/cs_flow/torch_model.py [hunk 1]
```diff
@@ -0,0 +1,531 @@
...
+        anomaly_map: Tensor
+        if mode == AnomalyMapMode.ALL:
+            anomaly_map = torch.ones(z_dists[0].shape[0], 1, *self.input_dims[1:]).to(z_dists[0].device)
+            for z_dist in z_dists:
+                anomaly_map *= F.interpolate(
+                    (z_dist**2).mean(dim=1, keepdim=True),
+                    size=self.input_dims[1:],
+                    mode="bilinear",
+                    align_corners=False,
+                )
```
Comment (djdameln) [inline anomalib/models/cs_flow/torch_model.py:462]: It would improve readability if you take the mean computation outside of the interpolate function, so something like:
```suggestion
                mean_z = (z_dist**2).mean(dim=1, keepdim=True)
                anomaly_map *= F.interpolate(
                    mean_z,
                    size=self.input_dims[1:],
                    mode="bilinear",
                    align_corners=False,
                )
```

File: anomalib/models/cs_flow/torch_model.py [hunk 2]
```diff
@@ -0,0 +1,531 @@
...
+        else:
+            anomaly_map = F.interpolate(
+                (z_dists[0] ** 2).mean(dim=1, keepdim=True),
+                size=self.input_dims[1:],
+                mode="bilinear",
+                align_corners=False,
+            )
```
Comment (djdameln) [inline anomalib/models/cs_flow/torch_model.py:469]:
```suggestion
            mean_z = (z_dists[0] ** 2).mean(dim=1, keepdim=True)
            anomaly_map = F.interpolate(
                mean_z,
                size=self.input_dims[1:],
                mode="bilinear",
                align_corners=False,
            )
```

File: anomalib/models/cs_flow/torch_model.py [hunk 3]
```diff
@@ -0,0 +1,531 @@
...
+        super().__init__()
+        self.input_dims = (num_channels, *input_size)
+        self.n_coupling_blocks = n_coupling_blocks
+        self.kernel_sizes = [3] * (n_coupling_blocks - 1) + [5]
+        self.clamp = clamp
+        self.cross_conv_hidden_channels = cross_conv_hidden_channels
+        self.feature_extractor = MultiScaleFeatureExtractor(n_scales=3, input_size=input_size)
+        self.graph = self._create_graph()
+
+    def _create_graph(self):
```
Comment (djdameln) [inline anomalib/models/cs_flow/torch_model.py:383]: Maybe you could place this stuff in a separate class which inherits from `GraphINN`, and which is instantiated from the constructor of the torch model. This would keep the `CsFlowModel` class a bit cleaner.

**Review by djdameln — CHANGES_REQUESTED** [review]
Thanks a lot for diving into this algo and adding the implementation. I have a few comments concerning the readability, but overall the PR looks good!

File: anomalib/models/cs_flow/torch_model.py [hunk 4]
```diff
@@ -0,0 +1,531 @@
...
+        perm = np.random.permutation(self.in_channels[index])
+        perm_inv = np.zeros_like(perm)
+        for idx, permutation in enumerate(perm):
+            perm_inv[permutation] = idx
+
+        perm = torch.LongTensor(perm)
+        perm_inv = torch.LongTensor(perm_inv)
+        return perm, perm_inv
+
+    # pylint: disable=unused-argument
```
[the remark this answers is not in the document]
Comment (ashwinvaidya17) [reply anomalib/models/cs_flow/torch_model.py:217]: `GraphINN` calls each submodule using keyword arguments.

File: anomalib/models/cs_flow/config.yaml [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+  name: mvtec #options: [mvtec, btech, folder]
+  format: mvtec
+  path: ./datasets/MVTec
+  category: bottle
+  task: segmentation
+  image_size: 768 # 768 is the dimensions used in the official implementation
+  train_batch_size: 16
+  test_batch_size: 16
+  inference_batch_size: 16
+  num_workers: 36
```
[the remark this answers is not in the document]
Comment (djdameln) [reply anomalib/models/cs_flow/config.yaml:11]: Yeah, let's make it 8 for all models. I believe for most models it's 8 already

**Comment by ashwinvaidya17** [comment]
I think I've addressed all the comments.

File: anomalib/models/csflow/torch_model.py [hunk 6]
```diff
@@ -0,0 +1,530 @@
...
+
+        out0 = out0 + y1_up
+        out1 = out1 + y0_down + y2_up
+        out2 = out2 + y1_down
+
+        if self.use_gamma:
+            out0 = out0 * self.gamma0
+            out1 = out1 * self.gamma1
+            out2 = out2 * self.gamma2
+        return out0, out1, out2
```
Comment (djdameln) [inline anomalib/models/csflow/torch_model.py:167]: This foward method is a bit hard to follow. Do you think it could be simplified by grouping some of the computations? For example, the Conv2D -> LeakyReLU -> Conv2D could be grouped into a sequential block or nn.Module class maybe? Otherwise it would probably suffice to add some comments in the `__init__` and `__forward__` methods.

**Review by djdameln — COMMENTED** [review]
Thanks, few comments left, mainly minor

File: anomalib/models/csflow/torch_model.py [hunk 7]
```diff
@@ -0,0 +1,530 @@
...
+        n_scales (int): Number of scales for input image.
+        input_size (Tuple[int, int]): Size of input image.
+    """
+
+    def __init__(self, n_scales: int, input_size: Tuple[int, int]):
+        super().__init__()
+
+        self.n_scales = n_scales
+        self.input_size = input_size
+        self.feature_extractor = TorchFXFeatureExtractor(
```
[the remark this answers is not in the document]
Comment (ashwinvaidya17) [reply anomalib/models/csflow/torch_model.py:466]: The paper mentions efficientnet so I've kept only that backbone. Plus we cannot pass model weights as enums to the current feature extractor via `config.yaml`. That's part of the feature extractor PR.

File: anomalib/models/csflow/torch_model.py (hunk 6 above)
Comment (ashwinvaidya17) [reply anomalib/models/csflow/torch_model.py]: I have added some comments but the problem is that since this is cross-scale, it performs computation on multiple inputs and returns multiple outputs. I am not sure if we can wrap some subset of the blocks here in a sequential module. But this is basically what this block does.
image

**Review by djdameln — APPROVED** [review]
Sorry, forgot to approve this. But I'm happy with the changes, so let's get this merged
